File: src/main/java/com/jozufozu/quench/api/Attunement.java
package com.jozufozu.quench.api;

import com.jozufozu.quench.api.interactions.AttackInteraction;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTBase;
import net.minecraft.util.text.ITextComponent;
import net.minecraftforge.common.util.INBTSerializable;

public abstract class Attunement<NBT extends NBTBase> implements INBTSerializable<NBT>
{
    public Attunement() {}

    protected ItemStack appliedOn;

    public abstract ITextComponent getDisplay(boolean advanced);

    public abstract int getColorRGB();

    public Attunement amplify()
    {
        return this;
    }

    public void onUserUpdate(EntityLivingBase user) { }

    public void onUserAttackedByEntity(AttackInteraction attack) { }

    public void onUserAttackEntity(AttackInteraction attack) { }

    public void onUserDeath() { }
}

File: src/main/java/com/jozufozu/symphony/api/Attunement.kt
package com.jozufozu.symphony.api

import com.jozufozu.symphony.api.interactions.AttackInteraction
import net.minecraft.entity.LivingEntity
import net.minecraft.nbt.INBT
import net.minecraft.util.text.ITextComponent

abstract class Attunement() {

    constructor(nbt: INBT): this()

    abstract val colorRGB: Int

    abstract fun getDisplay(advanced: Boolean): ITextComponent

    abstract fun serializeNBT(): INBT

    open fun onUserUpdate(user: LivingEntity) {}

    open fun onUserAttackedByEntity(attack: AttackInteraction) {}

    open fun onUserAttackEntity(attack: AttackInteraction) {}

    open fun onUserDeath() {}
}
